nx/lib/nx/defn/graph.ex

```diff
@@ -63,9 +71,28 @@ defmodule Nx.Defn.Graph do
         b = cos a       f32
         d = add b, c    f32
       >
+
+  ### Splitting with accumulator
+
+  With the arity-2 version of `expr_split_fn`, the decision to split can be made based on the expression and the accumulator.
+  This allows for more complex decisions to be made, such as splitting every 3 operations as in the example below. One might want to
+  split the graph to avoid running out of memory, for example.
+
+      # Count operations and split every 3 operations
+      split_fn = fn _tensor, count ->
+        new_count = count + 1
+        {count > 0 and rem(new_count, 3) == 0, new_count}
+      end
+
+      stages = Nx.Defn.Graph.split(expr, 0, split_fn)
   """
   def split(expr, expr_split_fn) when is_function(expr_split_fn, 1) do
-    {chain, _, _} = __split__(expr, expr_split_fn)
+    {chain, _, _} = __split__(expr, nil, fn node, acc -> {expr_split_fn.(node), acc} end)
+    chain
+  end
+
+  def split(expr, initial_acc, expr_split_fn) when is_function(expr_split_fn, 2) do
```

Comment:
This is effectively a separate function. So it needs a separate `@doc "..."` bit. Which I think is good, we don't need to mix the examples in the doc above.